java-checks/src/main/java/org/sonar/java/checks/spring/SpelExpressionCheck.java

```diff
@@ -0,0 +1,246 @@
+      return Collections.emptyList();
+    }
+  }
+
+  private static boolean isSpringAnnotation(AnnotationTree annotation) {
+    return annotation.symbolType().fullyQualifiedName().startsWith(SPRING_PREFIX);
+  }
+
+  private void checkSpringAnnotationArguments(AnnotationTree annotation) {
+    annotation.arguments().stream()
+      .filter(argument -> argument.is(Tree.Kind.STRING_LITERAL))
+      .forEach(argument -> checkSpringAnnotationStringArgument((LiteralTree) argument));
+  }
+
+  private void checkSpringAnnotationStringArgument(LiteralTree argument) {
+    try {
+      var argValue = argument.value();
+      var argContents = argValue.substring(1, argValue.length() - 1);
+      parseStringContents(argContents);
+    } catch (SyntaxError e) {
+      reportIssue(argument, e);
+    }
+  }
+
+  private void reportIssue(LiteralTree stringToken, SyntaxError error) {
+    var tokenStart = Position.startOf(stringToken);
+    var textSpan = new AnalyzerMessage.TextSpan(
+      tokenStart.line(),
+      tokenStart.columnOffset() + error.rangeStart,
+      tokenStart.line(),
+      tokenStart.columnOffset() + error.rangeEnd
+    );
+
+    var analyzerMessage = new AnalyzerMessage(this, context.getInputFile(), textSpan, error.getMessage(), 0);
+    ((DefaultJavaFileScannerContext) context).reportIssue(analyzerMessage);
+  }
+
+  private static void parseStringContents(String content) {
+    var i = 0;
+    while (i < content.length()) {
+      var c = content.charAt(i);
+      switch (c) {
+        case '$':
+          i = parseDelimitersAndContents(content, i + 1, i + 1, SpelExpressionCheck::checkValidPropertyPlaceholder);
+          break;
+        case '#':
+          i = parseDelimitersAndContents(content, i + 1, i + 1, SpelExpressionCheck::checkValidSpelExpression);
+          break;
+        default:
+          i++;
+          break;
+      }
+    }
+  }
+
+  private static int parseDelimitersAndContents(
+    String value,
+    int startIndex,
+    int rangeStart,
+    ObjIntConsumer<String> parseContents) {
+    if (startIndex == value.length()) {
+      return startIndex;
+    }
+    var endIndex = parseDelimiterBraces(value, startIndex, rangeStart);
+    if (endIndex == startIndex) {
+      return endIndex;
+    }
+    var contents = value.substring(startIndex + 1, endIndex - 1);
+    parseContents.accept(contents, rangeStart);
+    return endIndex;
+  }
+
+  private static int parseDelimiterBraces(String value, int startIndex, int rangeStart) throws SyntaxError {
+    if (value.charAt(startIndex) != '{') {
+      return startIndex;
+    }
+    var i = startIndex + 1;
+    var openCount = 1;
+
+    while (i < value.length()) {
+      var c = value.charAt(i);
+      i++;
+      switch (c) {
+        case '{':
+          openCount++;
+          break;
+        case '}':
+          openCount--;
+          if (openCount == 0) {
+            return i;
+          }
+          break;
+        default:
+          break;
+      }
+    }
```

Comment:
This could be a for loop as well
```suggestion
    var i = startIndex + 1;

    for (int opencount = 1; i < value.length(); i++) {
      var c = value.charAt(i);
      if (c == '{') {
          openCount++;
      } else if (c == '}') {
          openCount--;
          if (openCount == 0) {
            return i;
          }
       }
    }
```

Reply:
Fixed.